File: frontend/lib/approvals.ts
/**
 * Approval queue — types, Persian vocabulary and API wrappers.
 *
 * Field names are lifted verbatim from `backend/app/schemas/approvals.py`
 * (ApprovalRead / ApprovalReadWithNames / ApprovalSummary). Renaming anything
 * here silently produces `undefined` in the table instead of a type error,
 * because the responses arrive as plain JSON.
 *
 * The write side is split the same way the backend splits it: a requester may
 * describe an action, a reviewer may only approve or reject. There is
 * deliberately no client type that can set `status` directly — that would make
 * the whole gate bypassable in one call.
 */

import { api } from "@/lib/api-client";

// ---------------------------------------------------------------- vocabularies

export type ApprovalActionType =
  | "publish_article"
  | "bulk_publish"
  | "bulk_delete_content"
  | "restructure_categories"
  | "change_site_settings"
  | "delete_website"
  | "ai_auto_publish"
  | "ai_bulk_rewrite"
  | "ai_keyword_campaign";

export type ApprovalStatus =
  | "pending"
  | "approved"
  | "rejected"
  | "cancelled"
  | "executed"
  | "failed";

export type ApprovalPriority = "low" | "normal" | "high" | "urgent";

export type ApprovalRiskLevel = "low" | "medium" | "high" | "critical";

export const APPROVAL_ACTION_LABELS_FA: Record<string, string> = {
  publish_article: "انتشار مقاله",
  bulk_publish: "انتشار گروهی",
  bulk_delete_content: "حذف گروهی محتوا",
  restructure_categories: "بازچینی دسته‌بندی‌ها",
  change_site_settings: "تغییر تنظیمات سایت",
  delete_website: "حذف وب‌سایت",
  ai_auto_publish: "انتشار خودکار توسط هوش مصنوعی",
  ai_bulk_rewrite: "بازنویسی گروهی توسط هوش مصنوعی",
  ai_keyword_campaign: "کمپین کلمات کلیدی هوش مصنوعی",
};

/** Mirrors APPROVAL_ACTION_CATEGORIES in the model, for grouping the queue. */
export const APPROVAL_ACTION_CATEGORIES_FA: Record<string, string> = {
  publish_article: "محتوا",
  bulk_publish: "محتوا",
  bulk_delete_content: "محتوا",
  restructure_categories: "ساختاری",
  change_site_settings: "ساختاری",
  delete_website: "ساختاری",
  ai_auto_publish: "هوش مصنوعی",
  ai_bulk_rewrite: "هوش مصنوعی",
  ai_keyword_campaign: "هوش مصنوعی",
};

export const APPROVAL_STATUS_LABELS_FA: Record<string, string> = {
  pending: "در انتظار بررسی",
  approved: "تأیید شده",
  rejected: "رد شده",
  cancelled: "لغو شده",
  executed: "اجرا شده",
  failed: "اجرا با خطا",
};

export const APPROVAL_PRIORITY_LABELS_FA: Record<string, string> = {
  low: "کم",
  normal: "معمولی",
  high: "زیاد",
  urgent: "فوری",
};

export const APPROVAL_RISK_LABELS_FA: Record<string, string> = {
  low: "کم",
  medium: "متوسط",
  high: "زیاد",
  critical: "بحرانی",
};

/**
 * Statuses a request can no longer move out of. Mirrors
 * APPROVAL_TERMINAL_STATUSES in the model — the UI hides the decide/cancel
 * buttons on these instead of letting the user post a call that will 409.
 */
export const APPROVAL_TERMINAL_STATUSES: string[] = [
  "rejected",
  "cancelled",
  "executed",
  "failed",
];

export function isApprovalTerminal(status: string | null | undefined): boolean {
  if (!status) return false;
  return APPROVAL_TERMINAL_STATUSES.includes(status);
}

// -------------------------------------------------------------------- entities

export interface ApprovalRequest {
  id: string;
  organization_id: string;
  website_id: string | null;
  action_type: ApprovalActionType | string;
  status: ApprovalStatus | string;
  priority: ApprovalPriority | string;
  title: string;
  description: string | null;
  requester_id: string;
  reviewer_id: string | null;
  payload: Record<string, any>;
  risk_level: ApprovalRiskLevel | string;
  affected_items_count: number;
  decided_by: string | null;
  decided_at: string | null;
  reviewer_comment: string | null;
  executed_at: string | null;
  execution_error: string | null;
  execution_result: Record<string, any> | null;
  expires_at: string | null;
  related_article_id: string | null;
  related_brief_id: string | null;
  created_at: string;
  updated_at: string;
}

/** The list endpoint resolves display names via outer joins; all are nullable. */
export interface ApprovalRequestWithNames extends ApprovalRequest {
  requester_name: string | null;
  requester_email: string | null;
  reviewer_name: string | null;
  decided_by_name: string | null;
  website_name: string | null;
}

export interface ApprovalSummary {
  pending: number;
  pending_urgent: number;
  pending_high_risk: number;
  approved_awaiting_execution: number;
  expiring_soon: number;
  by_action_type: Record<string, number>;
  by_priority: Record<string, number>;
}

export interface ApprovalExpireResult {
  expired: number;
}

// ------------------------------------------------------------------- requests

export interface ApprovalCreateRequest {
  action_type: ApprovalActionType | string;
  title: string;
  description?: string;
  website_id?: string;
  reviewer_id?: string;
  priority?: ApprovalPriority;
  risk_level?: ApprovalRiskLevel;
  affected_items_count?: number;
  payload?: Record<string, any>;
  expires_in_hours?: number;
  related_article_id?: string;
  related_brief_id?: string;
}

export interface ApprovalDecisionRequest {
  decision: "approved" | "rejected";
  reviewer_comment?: string;
}

export interface ApprovalCancelRequest {
  reason?: string;
}

// ----------------------------------------------------------------- formatting

/**
 * Human countdown to the deadline. Returns null when there is no deadline, so
 * the caller can omit the column rather than print a placeholder.
 */
export function formatExpiryFa(expiresAt: string | null | undefined): string | null {
  if (!expiresAt) return null;
  const target = new Date(expiresAt);
  if (Number.isNaN(target.getTime())) return null;

  const diffMs = target.getTime() - Date.now();
  if (diffMs <= 0) return "مهلت پایان یافته";

  const hours = Math.floor(diffMs / 3_600_000);
  if (hours < 1) {
    const minutes = Math.max(1, Math.floor(diffMs / 60_000));
    return `${minutes.toLocaleString("fa-IR")} دقیقه باقی مانده`;
  }
  if (hours < 24) return `${hours.toLocaleString("fa-IR")} ساعت باقی مانده`;
  const days = Math.floor(hours / 24);
  return `${days.toLocaleString("fa-IR")} روز باقی مانده`;
}

/** True when the deadline is inside the window the summary calls "expiring soon". */
export function isExpiringSoon(
  expiresAt: string | null | undefined,
  withinHours = 24
): boolean {
  if (!expiresAt) return false;
  const target = new Date(expiresAt);
  if (Number.isNaN(target.getTime())) return false;
  const diffMs = target.getTime() - Date.now();
  return diffMs > 0 && diffMs <= withinHours * 3_600_000;
}

// ---------------------------------------------------------------------- calls

export function listApprovals(params?: {
  website_id?: string;
  status?: string;
  action_type?: string;
  priority?: string;
  mine_only?: boolean;
  assigned_to_me?: boolean;
  limit?: number;
  offset?: number;
}) {
  const qs = new URLSearchParams();
  if (params?.website_id) qs.set("website_id", params.website_id);
  if (params?.status) qs.set("status", params.status);
  if (params?.action_type) qs.set("action_type", params.action_type);
  if (params?.priority) qs.set("priority", params.priority);
  if (params?.mine_only !== undefined) qs.set("mine_only", String(params.mine_only));
  if (params?.assigned_to_me !== undefined)
    qs.set("assigned_to_me", String(params.assigned_to_me));
  if (params?.limit !== undefined) qs.set("limit", String(params.limit));
  if (params?.offset !== undefined) qs.set("offset", String(params.offset));
  const qsStr = qs.toString();
  return api.get<ApprovalRequestWithNames[]>(`/approvals${qsStr ? `?${qsStr}` : ""}`);
}

export function getApprovalSummary() {
  return api.get<ApprovalSummary>("/approvals/summary");
}

export function getApproval(approvalId: string) {
  return api.get<ApprovalRequest>(`/approvals/${approvalId}`);
}

export function createApproval(body: ApprovalCreateRequest) {
  return api.post<ApprovalRequest>("/approvals", body);
}

export function decideApproval(approvalId: string, body: ApprovalDecisionRequest) {
  return api.post<ApprovalRequest>(`/approvals/${approvalId}/decide`, body);
}

export function cancelApproval(approvalId: string, body?: ApprovalCancelRequest) {
  return api.post<ApprovalRequest>(`/approvals/${approvalId}/cancel`, body || {});
}

export function expireStaleApprovals() {
  return api.post<ApprovalExpireResult>("/approvals/expire-stale", {});
}

File: backend/app/schemas/approvals.py
"""Schemas for the approval queue.

Write models are split by actor, because the two sides of an approval must not
be able to touch each other's fields:

* `ApprovalCreate` — the requester. May set what the action is and what it
  affects, but not `status`, `decided_by`, or any execution field. If a client
  could post `status="approved"` the whole gate would be bypassable in one call.
* `ApprovalDecision` — the reviewer. May only approve or reject and leave a
  comment. `payload` is deliberately absent: the thing being approved must be
  exactly the thing that was requested, otherwise the reviewer signs off on one
  action and a different one executes.

`payload` is validated for size rather than shape. Each action_type carries its
own contract and the executing service checks it; a schema-level union over
nine action types would have to change every time a tenth is added.
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.approvals import (
    APPROVAL_ACTION_TYPES,
    APPROVAL_PRIORITIES,
    APPROVAL_RISK_LEVELS,
    APPROVAL_STATUSES,
)


# ------------------------------------------------------------------- read model

class ApprovalRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    website_id: UUID | None = None
    action_type: str
    status: str
    priority: str
    title: str
    description: str | None = None
    requester_id: UUID
    reviewer_id: UUID | None = None
    payload: dict = Field(default_factory=dict)
    risk_level: str
    affected_items_count: int
    decided_by: UUID | None = None
    decided_at: datetime | None = None
    reviewer_comment: str | None = None
    executed_at: datetime | None = None
    execution_error: str | None = None
    execution_result: dict | None = None
    expires_at: datetime | None = None
    related_article_id: UUID | None = None
    related_brief_id: UUID | None = None
    created_at: datetime
    updated_at: datetime


class ApprovalReadWithNames(ApprovalRead):
    """List view rows, enriched with display names.

    The queue screen shows "who asked" and "who decided" on every row. Resolving
    those two ids per row in the frontend would be N+1 requests, so the service
    joins them once and fills these in.
    """
    requester_name: str | None = None
    requester_email: str | None = None
    reviewer_name: str | None = None
    decided_by_name: str | None = None
    website_name: str | None = None


# ------------------------------------------------------------- requester writes

class ApprovalCreate(BaseModel):
    """Queue a new approval request.

    Note what is *not* here: status, decided_*, executed_*, execution_result.
    Those belong to the reviewer and the executor respectively.
    """
    action_type: str
    title: str = Field(min_length=3, max_length=500)
    description: str | None = Field(default=None, max_length=5000)
    website_id: UUID | None = None
    reviewer_id: UUID | None = None
    priority: str = "normal"
    risk_level: str = "medium"
    affected_items_count: int = Field(default=1, ge=0, le=1_000_000)
    payload: dict = Field(default_factory=dict)
    expires_in_hours: int | None = Field(default=None, ge=1, le=8760)
    related_article_id: UUID | None = None
    related_brief_id: UUID | None = None

    @field_validator("action_type")
    @classmethod
    def _known_action(cls, v: str) -> str:
        if v not in APPROVAL_ACTION_TYPES:
            raise ValueError(
                f"action_type must be one of {sorted(APPROVAL_ACTION_TYPES)}"
            )
        return v

    @field_validator("priority")
    @classmethod
    def _known_priority(cls, v: str) -> str:
        if v not in APPROVAL_PRIORITIES:
            raise ValueError(f"priority must be one of {sorted(APPROVAL_PRIORITIES)}")
        return v

    @field_validator("risk_level")
    @classmethod
    def _known_risk(cls, v: str) -> str:
        if v not in APPROVAL_RISK_LEVELS:
            raise ValueError(f"risk_level must be one of {sorted(APPROVAL_RISK_LEVELS)}")
        return v

    @field_validator("payload")
    @classmethod
    def _bounded_payload(cls, v: dict) -> dict:
        # Guard against a client stuffing an article body (or worse) into a
        # JSONB column that every queue page load then has to read back.
        if len(v) > 50:
            raise ValueError("payload must not exceed 50 top-level keys")
        return v


# -------------------------------------------------------------- reviewer writes

class ApprovalDecision(BaseModel):
    """Approve or reject a pending request.

    `decision` is restricted to the two reviewer-reachable outcomes. "cancelled"
    is the requester's verb and has its own endpoint; "executed"/"failed" are
    written by the executor and are never client-settable.
    """
    decision: str
    reviewer_comment: str | None = Field(default=None, max_length=2000)

    @field_validator("decision")
    @classmethod
    def _reviewer_decision_only(cls, v: str) -> str:
        allowed = {"approved", "rejected"}
        if v not in allowed:
            raise ValueError(f"decision must be one of {sorted(allowed)}")
        return v


class ApprovalCancel(BaseModel):
    """Requester withdraws their own pending request."""
    reason: str | None = Field(default=None, max_length=1000)


# ---------------------------------------------------------------- aggregate views

class ApprovalSummary(BaseModel):
    """Counts for the queue badge and the dashboard card."""
    pending: int = 0
    pending_urgent: int = 0
    pending_high_risk: int = 0
    approved_awaiting_execution: int = 0
    expiring_soon: int = 0
    by_action_type: dict[str, int] = Field(default_factory=dict)
    by_priority: dict[str, int] = Field(default_factory=dict)


class ApprovalExpireResult(BaseModel):
    """Outcome of the periodic sweep over overdue requests."""
    expired: int = 0


__all__ = [
    "ApprovalRead",
    "ApprovalReadWithNames",
    "ApprovalCreate",
    "ApprovalDecision",
    "ApprovalCancel",
    "ApprovalSummary",
    "ApprovalExpireResult",
    "APPROVAL_ACTION_TYPES",
    "APPROVAL_STATUSES",
    "APPROVAL_PRIORITIES",
    "APPROVAL_RISK_LEVELS",
]
